src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VsSolutionUpdatesProjectSnapshotChangeTrigger.cs

```diff
@@ -43,22 +46,33 @@ internal class VsSolutionUpdatesProjectSnapshotChangeTrigger : ProjectSnapshotCh
                 throw new ArgumentNullException(nameof(workspaceStateGenerator));
             }
 
+            if (joinableTaskContext is null)
+            {
+                throw new ArgumentNullException(nameof(joinableTaskContext));
+            }
+
             _services = services;
             _projectService = projectService;
             _workspaceStateGenerator = workspaceStateGenerator;
+            _joinableTaskContext = joinableTaskContext;
         }
 
         public override void Initialize(ProjectSnapshotManagerBase projectManager)
         {
             _projectManager = projectManager;
 
-            // Attach the event sink to solution update events.
-            if (_services.GetService(typeof(SVsSolutionBuildManager)) is IVsSolutionBuildManager solutionBuildManager)
+            _ = _joinableTaskContext.Factory.RunAsync(async () =>
             {
-                // We expect this to be called only once. So we don't need to Unadvise.
-                var hr = solutionBuildManager.AdviseUpdateSolutionEvents(this, out _);
-                Marshal.ThrowExceptionForHR(hr);
-            }
+                await _joinableTaskContext.Factory.SwitchToMainThreadAsync();
```

Comment:
Given [this](https://github.com/dotnet/aspnetcore-tooling/pull/4046) change works without this PR the work here is purely perf based right?

Reply:
no, I think it is to fix a thread rule violation problem, because Initialize is called on the thread pool, and this leads a random product hang during closing the solution.

The question here is that the new code never ensures the solution build event is hooked up when Initialize is finished. It is logically possible that the event is not chained when a build is finished, so the code may miss the event, although it is fairly unlikely in the real product, because build typically happens in the background, and the UI thread does pump message for a while, unless up to date check is so efficient, so the build essentially is skipped. By looking this alone, we won't be able to say whether this would lead a problem here.

Reply:
Yes, this change was just to fix the threading rule violation.